### PR TITLE
Listing API

### PR DESCRIPTION
**Pull request by zhou-michael**
POST /api/listing
GET /api/listing/[listing_id]
PATCH /api/listing/[listing_id]

PATCH is callable regardless of user calling, possibly will later implement can only update title/description etc if user_id matches owner_id, and can only update buyer_list if user_id doesn't match owner_id

tested with postman

pls review

### CODE REVIEW

File: src/app/api/listing/[listing_id]/route.ts
Comment (JonathonYY) [inline src/app/api/listing/[listing_id]/route.ts:1]: Use `const` instead of `let` when not reassigning

`GET` API shouldn't return all fields of the listing - see API spec
`GET` API also does not currently handle errors properly; just have `getListing` throw an error from `getListing` and catch it in the API (see comments on `getListing`)

E.g. instead of:
```
let { result, error } = await getListing(listing_id)
return NextResponse.json({ data: result, error: error });
```
do
```
const result = await getListing(listing_id);
return NextResponse.json({ data: result, error: null });
```

`PATCH`: Same comment on error handling as `GET`
Add checking to ensure only valid fields are passed in for update (see Edwin's `PATCH` user for an example)

Remove `newListing` from imports as it isn't used anymore

File: src/app/api/listing/route.ts
Comment (JonathonYY) [inline src/app/api/listing/route.ts:1]: Should have `const { result, error }` instead of `let { result, error }`

Remove `TODO` comment on `POST`
Add manual error checking for missing arguments instead of leaving it to be `"code": "invalid-argument", "name": "FirebaseError"` — see Edwin's `POST` API for user

File: src/lib/firebase/firestore/listing/addListing.ts
Comment (JonathonYY) [inline src/lib/firebase/firestore/listing/addListing.ts:1]: Use `const` instead of `let` when you aren't reassigning

Add new listing to owner's `active_listings`

File: src/lib/firebase/firestore/listing/getListing.ts
Comment (JonathonYY) [inline src/lib/firebase/firestore/listing/getListing.ts:1]: This currently doesn't handle errors with missing listing (which *should* have been found via Postman)

Don't catch error here, that's why we wrap the helper in a `try-catch` block in the API itself

This fixes the API:
```
export default async function getListing(doc_id: string) {
    const docRef = doc(db, "listings", doc_id);
    const docSnapshot = await getDoc(docRef);

    if (docSnapshot.exists()) {
        return docSnapshot.data();
    } else {
        throw new Error("No listing exists for given id");
    }
}
```

File: src/lib/firebase/firestore/listing/patchListing.ts
Comment (JonathonYY) [inline src/lib/firebase/firestore/listing/patchListing.ts:1]: I think it's best to leave the thrown error so that the API itself catches it

**Review by JonathonYY — CHANGES_REQUESTED** [review]
General comments:
Stylistically, use `const` instead of `let` when possible, and remember `;` at end of lines

Add unit tests for your APIs

File: jest.config.js
Comment (JonathonYY) [inline jest.config.js:1]: This file should be terminated with a newline

File: src/lib/firebase/firestore/types.ts
Comment (JonathonYY) [inline src/lib/firebase/firestore/types.ts:1]: Watch out for merge conflicts in this file (different order of fields), `User` changes

File: src/lib/firebase/firestore/listing/patchListing.ts
Comment (JonathonYY) [inline src/lib/firebase/firestore/listing/patchListing.ts:1]: PATCH listing API shouldn't include `reporters` or `ratings` in the return `data` - either clean it here or in the API itself, should be able to just use `delete`

Should be `const update_data` as it isn't reassigned

File should be terminated with a newline

File: src/lib/firebase/firestore/listing/getListing.ts
Comment (JonathonYY) [inline src/lib/firebase/firestore/listing/getListing.ts:1]: Instead of using a separate `result` and `forEach` to create the output, you can just do:
```
delete data.ratings;
delete data.reporters;
```
And return `data`

This should work to remove the fields that shouldn't be returned

Also, file should be newline terminated

File: src/lib/firebase/firestore/listing/addListing.ts
Comment (JonathonYY) [reply src/lib/firebase/firestore/listing/addListing.ts:1]: `error` is never used, you can remove it
You should just have `const result = ...` at the end, instead of `let result` at the start

`listing_data` should be const
`seller_rating` should be `user_data.cum_seller_rating/user_data.completed_sales`; be sure to account for potential divide by 0 when there are no completed sales
Within `listing_data`, can you ensure the text fields used for querying (`condition`, `category`, `data.title`, `owner_name`) are lower-case so querying works properly?

Once again, the new listing **must** be added to the `active_listings` of the `owner`

File should be newline terminated

File: src/app/api/listing/[listing_id]/route.ts
Comment (JonathonYY) [inline src/app/api/listing/[listing_id]/route.ts:1]: Instead of `let result` use `const result`

Either here or in `patchListing`, add something to ensure searchable text fields are lowercase (see comment on `addListing`)

File: src/app/api/listing/[listing_id]/route.test.ts
Comment (JonathonYY) [inline src/app/api/listing/[listing_id]/route.test.ts:1]: `GET` and `PATCH` should be given separate `describe` blocks as they're different APIs

File: src/app/api/listing/route.test.ts
Comment (JonathonYY) [inline src/app/api/listing/route.test.ts:1]: As you've mocked out `db`, confirm that `POST` is adding expected values to it
This is important because there are some fields that shouldn't be returned by `GET` but that should still end up in `db`

File: src/lib/firebase/firestore/listing/addListing.test.ts
Comment (JonathonYY) [inline src/lib/firebase/firestore/listing/addListing.test.ts:1]: Once fixed, add a test to ensure division by zero to get `seller_rating` is handled properly

**Review by JonathonYY — CHANGES_REQUESTED** [review]
Overall APIs & tests look pretty good

Couple code style nits, and a few requests for proper functionality/to make querying possible

File: jest.config.js
Comment (zhou-michael) [reply jest.config.js:1]: fixed

File: src/lib/firebase/firestore/listing/getListing.ts
Comment (zhou-michael) [reply src/lib/firebase/firestore/listing/getListing.ts:1]: thanks, fixed this

File: src/app/api/listing/[listing_id]/route.ts
Comment (zhou-michael) [reply src/app/api/listing/[listing_id]/route.ts:1]: should be done !

File: src/app/api/listing/[listing_id]/route.test.ts
Comment (zhou-michael) [reply src/app/api/listing/[listing_id]/route.test.ts:1]: done as well

File: src/lib/firebase/firestore/listing/addListing.test.ts
Comment (zhou-michael) [reply src/lib/firebase/firestore/listing/addListing.test.ts:1]: done as well, seller_rating currently set to 0 if zero completed sales

File: src/app/api/listing/route.test.ts
Comment (zhou-michael) [reply src/app/api/listing/route.test.ts:1]: done

File: src/lib/firebase/firestore/listing/patchListing.ts
Comment (JonathonYY) [reply src/lib/firebase/firestore/listing/patchListing.ts:1]: Same as prior comment - return shouldn't include `reporters` or `rating`

Apply the same fix as you did in`src/lib/firebase/firestore/listing/getListing.ts`

File: src/app/api/listing/[listing_id]/route.test.ts
Comment (JonathonYY) [reply src/app/api/listing/[listing_id]/route.test.ts:1]: Overall looks good, but remember to update the `PATCH` tests once the return contents are fixed

File: src/app/api/listing/[listing_id]/route.ts
Comment (JonathonYY) [inline src/app/api/listing/[listing_id]/route.ts:1]: Remove `newListing` as it isn't used
Otherwise, looks good

File: src/app/api/listing/route.test.ts [hunk 1]
```diff
@@ -37,3 +74,241 @@ describe('Test GET all listings API endpoint', () => {
...
+
+        const jsonResponse = await response.json();
+
+        expect(doc.mock.calls[0][1]).toBe('users')
+        expect(doc.mock.calls[0][2]).toBe('user_invalid')
+        expect(getDoc).toHaveBeenCalled();
+        expect(addDoc).not.toHaveBeenCalled();
+
+        // check for correct output
+        expect(jsonResponse.data).toBeNull;
```
Comment (JonathonYY) [inline src/app/api/listing/route.test.ts]: Should be `toBeNull()`

File: src/app/api/listing/route.test.ts [hunk 2]
```diff
@@ -1,8 +1,45 @@
 import { GET } from "./route";
+import { POST } from './route';
+import { GET as GET2 } from "./[listing_id]/route";
 import * as getAllListings from "@/lib/firebase/firestore/listing/getAllListings";
+import { NextResponse } from 'next/server';
+import { ACTION_SERVER_ACTION } from "next/dist/client/components/router-reducer/router-reducer-types";
```
Comment (JonathonYY) [inline src/app/api/listing/route.test.ts]: I'm pretty sure you can remove `import { ACTION_SERVER_ACTION }` unless there's something I'm missing

File: src/lib/firebase/firestore/listing/addListing.ts [hunk 3]
```diff
@@ -1,21 +1,46 @@
...
-}
-
-export default async function addListing(data: DataToInsert) {
-  let result = null;
-  let error = null;
+export default async function addListing(data: AddListingData) {
+  // get user from user_id
+  const userRef = doc(db, "users", data.user_id);
+  const userSnapshot = await getDoc(userRef);
+  let user_data: User
```
Comment (JonathonYY) [inline src/lib/firebase/firestore/listing/addListing.ts]: Add `;`

File: src/lib/firebase/firestore/listing/addListing.ts [hunk 4]
```diff
@@ -1,21 +1,46 @@
...
+    potential_buyers: [], // user_ids of potential buyers
+    reporters: [], // user_ids of reporters
+    ratings: {}, // strings are user_ids mapped to number ratings
+    image_paths: data.image_paths, // list of paths to imgs
+  };
+
+  const docRef = await addDoc(collection(db, "listings"), listing_data);
+  const listing_id = docRef.id;
+
+  updateUser(data.user_id, { active_listings: user_data.active_listings.push(listing_id) });
```
Comment (JonathonYY) [inline src/lib/firebase/firestore/listing/addListing.ts]: You should use `arrayUnion()` instead here

https://firebase.google.com/docs/firestore/manage-data/add-data#update_elements_in_an_array

File: src/lib/firebase/firestore/listing/patchListing.test.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+            'category': 'food',
+            'description': '',
+            'owner': 'user1',
+            'owner_name': 'A',
+            'owner_pfp': '',
+            'seller_rating': 4,
+            'selected_buyer': '',
+            'potential_buyers': [],
+            'reporters': [],
+            'ratings': {},
```
Comment (JonathonYY) [inline src/lib/firebase/firestore/listing/patchListing.test.ts]: Update test to reflect that `reporters` and `ratings` shouldn't be returned

**Review by JonathonYY — CHANGES_REQUESTED** [review]
A couple more changes to match API spec and a few nits, but overall I think it looks good

File: src/lib/firebase/firestore/listing/patchListing.test.ts (hunk 5 above)
Comment (zhou-michael) [reply src/lib/firebase/firestore/listing/patchListing.test.ts]: okie all of these should be addressed in latest commit !

File: src/lib/firebase/firestore/user/userUtil.ts [hunk 6]
```diff
@@ -32,6 +32,8 @@ export async function getUser(user_id: string): Promise<User> {
 export async function updateUser(user_id: string, data: { [key: string]: any }): Promise<User> {
   // get user to check if it exists
   await getUser(user_id);
+  console.log(user_id);
+  console.log(data);
```
Comment (JonathonYY) [inline src/lib/firebase/firestore/user/userUtil.ts]: Probably shouldn't include these log statements when we merge

File: src/app/api/listing/route.ts [hunk 7]
```diff
@@ -1,4 +1,6 @@
 import { NextResponse } from "next/server";
+import { Listing, newListing, AddListingData } from "@/lib/firebase/firestore/types";
```
Comment (Edwinhe03) [inline src/app/api/listing/route.ts]: remove unnecessary imports

File: src/lib/firebase/firestore/listing/addListing.ts [hunk 8]
```diff
@@ -1,21 +1,47 @@
 import { db } from "../../config";
-import { collection, addDoc } from "firebase/firestore";
+import { doc, collection, getDoc, addDoc, updateDoc, serverTimestamp, Timestamp, arrayUnion } from "firebase/firestore";
+import { User, Listing, AddListingData } from "../types";
+import { updateUser } from "../user/userUtil";
```
Comment (Edwinhe03) [inline src/lib/firebase/firestore/listing/addListing.ts]: remove unnecessary import

File: src/lib/firebase/firestore/listing/addListing.ts [hunk 9]
```diff
@@ -1,21 +1,47 @@
...
-}
-
-export default async function addListing(data: DataToInsert) {
-  let result = null;
-  let error = null;
+export default async function addListing(data: AddListingData) {
+  // get user from user_id
+  const userRef = doc(db, "users", data.user_id);
+  const userSnapshot = await getDoc(userRef);
+  let user_data: User;
```
Comment (Edwinhe03) [inline src/lib/firebase/firestore/listing/addListing.ts]: avoid using `let` in general. throw error first, then assign `const`.

```
if (!userSnapshot.exists()) {
  throw new Error("No user exists for given id");
}
const user_data: User = userSnapshot.data() as User;
```

File: src/app/api/listing/route.ts [hunk 10]
```diff
@@ -18,12 +20,44 @@ import getAllListings from "@/lib/firebase/firestore/listing/getAllListings";
...
+    ];
+
+    // validate input for only valid fields
+    Object.keys(data).forEach((key) => {
+      if (!desired_fields.includes(key)) {
+        throw new Error('invalid listing field');
+      }
+    });
+
+    // validate all fields are present
```
Comment (Edwinhe03) [inline src/app/api/listing/route.ts:45]: can we check that price is non-negative?

File: src/lib/firebase/firestore/listing/addListing.ts [hunk 11]
```diff
@@ -1,21 +1,47 @@
...
+    price: data.price,
+    condition: data.condition.toLowerCase(),
+    category: data.category.toLowerCase(),
+    description: data.description,
+    owner: data.user_id, // owner (seller) user_id
+    owner_name: `${user_data.first} ${user_data.last}`.toLowerCase(),
+    owner_pfp: user_data.pfp,
+    seller_rating: user_data.completed_sales
+                    ? user_data.cum_seller_rating / user_data.completed_sales
+                    : 0,
```
Comment (Edwinhe03) [inline src/lib/firebase/firestore/listing/addListing.ts]: hm... not sure if we want to default to 0 for no ratings. maybe `-1` then have frontend handle as special case? @JonathonYY

File: src/app/api/listing/route.test.ts [hunk 12]
```diff
@@ -37,3 +74,241 @@ describe('Test GET all listings API endpoint', () => {
...
+        const jsonResponse = await response.json();
+
+        expect(doc.mock.calls[0][1]).toBe('users')
+        expect(doc.mock.calls[0][2]).toBe('user_invalid')
+        expect(getDoc).toHaveBeenCalled();
+        expect(addDoc).not.toHaveBeenCalled();
+
+        // check for correct output
+        expect(jsonResponse.data).toBeNull();
+        expect(jsonResponse.error).not.toBeNull();
```
Comment (Edwinhe03) [inline src/app/api/listing/route.test.ts]: you should expect the exact error message `"invalid listing field"`. same for all your other test cases. also add one for error on negative price

File: src/app/api/listing/[listing_id]/route.ts [hunk 13]
```diff
@@ -17,12 +19,19 @@ export async function GET(
...
+  } catch (e: unknown) {
+    if (e instanceof Error) {
+      return NextResponse.json({ data: null, error: e.message });
+    } else {
+      return NextResponse.json({ data: null, error: "unknown error" });
+    }
+  }
 }
 
 /*
```
Comment (Edwinhe03) [inline src/app/api/listing/[listing_id]/route.ts:37]: can you update docstring to have `?` after every param to indicate it is optional. ie. `title?: ...` (except `listing_id` which is required

File: src/app/api/listing/[listing_id]/route.test.ts [hunk 14]
```diff
@@ -13,6 +16,305 @@ const deleteListingMock = jest.spyOn(deleteListing, "default").mockImplementatio
...
+        const response: NextResponse = await GET(mockReq, { params: mockParams });
+
+        const jsonResponse = await response.json();
+
+        expect(doc.mock.calls[0][1]).toBe('listings')
+        expect(doc.mock.calls[0][2]).toBe('invalid_id')
+
+        // check for correct output
+        expect(jsonResponse.data).toBeNull();
+        expect(jsonResponse.error).not.toBeNull();
```
Comment (Edwinhe03) [inline src/app/api/listing/[listing_id]/route.test.ts]: ditto on specific error checks

File: src/lib/firebase/firestore/listing/getListing.ts [hunk 15]
```diff
@@ -0,0 +1,18 @@
+import { db } from "../../config";
+import { doc, getDoc } from "firebase/firestore";
+
+export default async function getListing(doc_id: string) {
+    const docRef = doc(db, "listings", doc_id);
+    const docSnapshot = await getDoc(docRef);
+
+    if (docSnapshot.exists()) {
+        let result = docSnapshot.data();
```
Comment (Edwinhe03) [inline src/lib/firebase/firestore/listing/getListing.ts]: i think you can use `const` here even though we delete fields

File: src/lib/firebase/firestore/listing/patchListing.ts [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+    });
+
+    const docSnapshot = await getDoc(docRef);
+
+    if (docSnapshot.exists()) {
+        // update each entry
+        await updateDoc(docRef, update_data)
+
+        const docSnapshot2 = await getDoc(docRef);
+        if (docSnapshot2.exists()) {
```
Comment (Edwinhe03) [inline src/lib/firebase/firestore/listing/patchListing.ts]: can u move lines 25-34 to below the outer if statement to reduce nesting?

File: src/lib/firebase/firestore/listing/patchListing.test.ts [hunk 17]
```diff
@@ -0,0 +1,118 @@
...
+            'image_paths': [], 
+        });
+    });
+
+    it('patchListing with invalid doc_id', async () => {
+        expect(async () => {
+            await patchListing('blah', { 'title': 'adsf' });
+        }).rejects.toThrow('No listing exists for given id');
+    });
+});
```
Comment (Edwinhe03) [inline src/lib/firebase/firestore/listing/patchListing.test.ts:118]: add test for invalid field

**Review by Edwinhe03 — CHANGES_REQUESTED** [review]
a bunch of nits but looks good after that. need sign off from someone else on default rating value if user has no prior ratings

File: src/lib/firebase/firestore/listing/addListing.ts (hunk 11 above)
Comment (zhou-michael) [reply src/lib/firebase/firestore/listing/addListing.ts]: setting to 3.5

File: src/app/api/listing/route.ts (hunk 10 above)
Comment (zhou-michael) [reply src/app/api/listing/route.ts:45]: done

File: src/lib/firebase/firestore/listing/addListing.ts (hunk 9 above)
Comment (zhou-michael) [reply src/lib/firebase/firestore/listing/addListing.ts]: fixed

File: src/app/api/listing/[listing_id]/route.ts (hunk 13 above)
Comment (zhou-michael) [reply src/app/api/listing/[listing_id]/route.ts:37]: gotchu

File: src/lib/firebase/firestore/listing/getListing.ts (hunk 15 above)
Comment (zhou-michael) [reply src/lib/firebase/firestore/listing/getListing.ts]: oh cool okie

File: src/lib/firebase/firestore/listing/patchListing.ts (hunk 16 above)
Comment (zhou-michael) [reply src/lib/firebase/firestore/listing/patchListing.ts]: fixed

File: src/app/api/listing/route.test.ts (hunk 12 above)
Comment (zhou-michael) [reply src/app/api/listing/route.test.ts]: done ! also implemented in PATCH + included test

File: src/lib/firebase/firestore/listing/patchListing.test.ts (hunk 17 above)
Comment (zhou-michael) [reply src/lib/firebase/firestore/listing/patchListing.test.ts:118]: i believe i catch invalid fields in the actual API route and also test it there

Comment (Edwinhe03) [reply src/lib/firebase/firestore/listing/patchListing.test.ts:118]: ah yep, you are right

File: src/lib/firebase/firestore/types.ts [hunk 18]
```diff
@@ -1,5 +1,26 @@
...
+  user_id: string,
+  title: string,
+  price: number,
+  condition: string,
+  category: string,
+  description: string,
+  image_paths: string[], // list of paths to imgs
+}
+
+export interface PatchListingData {
```
Comment (Edwinhe03) [inline src/lib/firebase/firestore/types.ts:13]: sorry forgot to say these should also be optional (with ?) but let's leave it for a follow up/clean up pr

Comment (zhou-michael) [reply src/lib/firebase/firestore/types.ts:13]: oki

**Review by Edwinhe03 — APPROVED** [review]
lgtm

**Review by JonathonYY — APPROVED** [review]
LGTM
